File: data_structures_and_algorithms/Code_401/class-15/solutions/python/binary-tree/binary_tree.py
from collections import deque


class BinaryTree:
    def __init__(self, root=None, values=None):
        self.root = root
        if values:
            for value in values:
                self.add(value)

    def pre_order(self, action=None):
        def walk(root, action):

            if not root:
                return

            action(root.value)
            walk(root.left, action)
            walk(root.right, action)

        walk(self.root, action or print)

    def in_order(self, action=None):
        def walk(root, action):

            if not root:
                return

            walk(root.left, action)
            action(root.value)
            walk(root.right, action)

        walk(self.root, action or print)

    def post_order(self, action=None):
        def walk(root, action):

            if not root:
                return

            walk(root.left, action)
            walk(root.right, action)
            action(root.value)

        walk(self.root, action or print)

    def breadth_first(self, action=None):

        if not self.root:
            return

        action = action or print

        breadth = Queue()

        breadth.enqueue(self.root)

        while not breadth.is_empty():
            front = breadth.dequeue()

            action(front.value)

            if front.left:
                breadth.enqueue(front.left)

            if front.right:
                breadth.enqueue(front.right)

    def add(self, value):

        node = Node(value)

        if not self.root:
            self.root = node
            return

        breadth = Queue()

        breadth.enqueue(self.root)

        while not breadth.is_empty():
            front = breadth.dequeue()
            if not front.left:
                front.left = node
                return
            else:
                breadth.enqueue(front.left)

            if not front.right:
                front.right = node
                return
            else:
                breadth.enqueue(front.right)

    def find_maximum_value(self):

        max = self.root.value

        def max_check(value):
            nonlocal max
            max = value if value > max else max

        self.pre_order(max_check)

        return max


class BinarySearchTree(BinaryTree):
    def add(self, value):
        def walk(root, node_to_add):

            if not root:
                return

            if node_to_add.value < root.value:
                if root.left:
                    walk(root.left, node_to_add)
                else:
                    root.left = node_to_add
            else:
                if root.right:
                    walk(root.right, node_to_add)
                else:
                    root.right = node_to_add

        node = Node(value)

        if not self.root:
            self.root = node
            return

        walk(self.root, node)

    def contains(self, value):
        def walk(value, root):

            if not root:
                return False

            return (
                root.value == value or walk(value, root.left) or walk(value, root.right)
            )

        return walk(value, self.root)


class Node:
    def __init__(self, value, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right


class Queue:
    def __init__(self):
        self.storage = deque()

    def enqueue(self, value):
        self.storage.appendleft(value)

    def dequeue(self):
        return self.storage.pop()

    def peek(self):
        return storage[-1]

    def is_empty(self):
        return len(self.storage) == 0

File: data_structures_and_algorithms/Code_401/class-16/solutions/java/BinaryTree.java
import java.util.LinkedList;
import java.util.ArrayList;
import java.util.Queue;

class BinaryTree {
  Node root;

  public int findMaximumValue() throws Exception{
    if(root == null) throw new Exception("no values to compare");
    int max = root.value;
    Queue<Node> q = new LinkedList<>();
    q.add(root);

    while(!q.isEmpty()){
      Node curr = q.poll();
      if(curr.value > max) max = curr.value;
      if(curr.left != null) q.add(curr.left);
      if(curr.right != null) q.add(curr.right);
    }
    return max;
  }

  public ArrayList<Integer> preorder(){
    ArrayList<Integer> vals = new ArrayList<>();
    preorder(root, vals);
    return vals;
  }
  public void preorder(Node current, ArrayList<Integer> vals){
    if(current == null) return;
    vals.add(current.value);
    preorder(root.left, vals);
    preorder(root.right, vals);
  }

  public ArrayList<Integer> inOrder(){
    ArrayList<Integer> vals = new ArrayList<>();
    inOrder(root, vals);
    return vals;
  }
  public void inOrder(Node current, ArrayList<Integer> vals){
    if(current == null) return;
    inOrder(root.left, vals);
    vals.add(current.value);
    inOrder(root.right, vals);
  }

  public ArrayList<Integer> postOrder(){
    ArrayList<Integer> vals = new ArrayList<>();
    postOrder(root, vals);
    return vals;
  }
  public void postOrder(Node current, ArrayList<Integer> vals){
    if(current == null) return;
    postOrder(root.left, vals);
    postOrder(root.right, vals);
    vals.add(current.value);
  }

  public String toString(){
    StringBuilder out = new StringBuilder();
    Queue<Node> q = new LinkedList<>();
    Queue<Node> q2 = new LinkedList<>();
    q.add(root);

    while(!q.isEmpty()){
      Node curr = q.poll();
      out.append(curr.value + ",");
      if(curr.left != null) q2.add(curr.left);
      if(curr.right != null) q2.add(curr.right);
      if(q.isEmpty()) {
        q = q2;
        q2 = new LinkedList<>();
        out.append("\n");
      }

    }
    return out.toString();
  }

  public static class Node {
    int value;
    Node left;
    Node right;

    public Node(int value){
      this.value = value;
    }
  }
}
